src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,578 @@
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.quota_transport = nullptr;
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+  std::unique_ptr<ServiceControlClient> client =
+      CreateServiceControlClient("unknown", kServiceConfigId, options);
+
+  client->Quota(quota_request1_, &quota_response,
+                [&done_status](Status status) { done_status = status; });
+
+  EXPECT_EQ(done_status, Status(Code::INVALID_ARGUMENT, "transport is NULL."));
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 0);
+
+  Statistics stat;
+  Status stat_status = client->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 1);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 0);
+}
+
+// Cached: false, Callback: stored
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithStoredCallbackMultipleReqeusts) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillRepeatedly(
+          Invoke(&mock_quota_transport_,
+                 &MockQuotaTransport::AllocateQuotaWithStoredCallback));
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  // call Quota 10 times
+  for (int i = 0; i < 10; i++) {
+    noncached_client_->Quota(
+        quota_request1_, &quota_response,
+        [&done_status](Status status) { done_status = status; });
+    EXPECT_EQ(done_status, Status::UNKNOWN);
+  }
+
+  // count store callback
+  EXPECT_EQ(mock_quota_transport_.on_done_vector_.size(), 10);
+
+  // execute stored callback
+  for (auto callback : mock_quota_transport_.on_done_vector_) {
+    done_status = Status::UNKNOWN;
+    callback(done_status.OK);
+    EXPECT_EQ(done_status, Status::OK);
+  }
+
+  Statistics stat;
+  Status stat_status = noncached_client_->GetStatistics(&stat);
+
+  EXPECT_EQ(stat_status, Status::OK);
+  EXPECT_EQ(stat.total_called_quotas, 10);
+  EXPECT_EQ(stat.send_quotas_by_flush, 0);
+  EXPECT_EQ(stat.send_quotas_in_flight, 10);
+}
+
+// Cached: true, Callback: stored
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestCachedQuotaWithStoredCallbackMultipleRequests) {
+  EXPECT_CALL(mock_quota_transport_, Quota(_, _, _))
+      .WillOnce(Invoke(&mock_quota_transport_,
+                       &MockQuotaTransport::AllocateQuotaWithStoredCallback));
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  cached_client_->Quota(
+      quota_request1_, &quota_response,
+      [&done_status](Status status) { done_status = status; });
+  EXPECT_EQ(done_status, Status::UNKNOWN);
+
+  // call Quota 10 times
+  for (int i = 0; i < 10; i++) {
+    cached_client_->Quota(
+        quota_request1_, &quota_response,
+        [&done_status](Status status) { done_status = status; });
+    EXPECT_EQ(done_status, Status::OK);
```

Comment:
its better not to use same done_status variable here.  let the one defined in line 244 stay with unknown